src/libraries/System.Text.RegularExpressions/tests/System.Text.RegularExpressions.Generators.Tests/RegexGeneratorParserTests.cs

```diff
@@ -0,0 +1,507 @@
+        public async Task Valid_MultiplRegexMethodsPerClass()
+        {
+            Assert.Empty(await RunGenerator(@"
+                using System.Text.RegularExpressions;
+                partial class C1
+                {
+                    [RegexGenerator(""a"")]
+                    private static partial Regex A();
+
+                    [RegexGenerator(""b"")]
+                    public static partial Regex B();
+
+                    [RegexGenerator(""b"")]
+                    public static partial Regex C();
+                }
+                partial class C2
+                {
+                    [RegexGenerator(""d"")]
+                    public static partial Regex D();
+
+                    [RegexGenerator(""d"")]
+                    public static partial Regex E();
+                }
+            ", compile: true));
+        }
+
+        [Fact]
+        public async Task Valid_OnStruct()
+        {
+            Assert.Empty(await RunGenerator(@"
+                using System.Text.RegularExpressions;
+                internal partial struct C
+                {
+                    [RegexGenerator(""ab"")]
+                    private static partial Regex Valid();
+                }
+            ", compile: true));
+        }
+
+        [Fact]
+        public async Task Valid_OnRecord()
+        {
+            Assert.Empty(await RunGenerator(@"
+                using System.Text.RegularExpressions;
+                internal partial record C
+                {
+                    [RegexGenerator(""ab"")]
+                    private static partial Regex Valid();
+                }
+            ", compile: true));
+        }
+
+        [Fact]
+        public async Task Valid_OnRecordStruct()
+        {
+            Assert.Empty(await RunGenerator(@"
+                using System.Text.RegularExpressions;
+                internal partial record struct C
+                {
+                    [RegexGenerator(""ab"")]
+                    private static partial Regex Valid();
+                }
+            ", compile: true));
+        }
+
+        [Fact]
+        public async Task Valid_NestedVaryingTypes()
+        {
+            Assert.Empty(await RunGenerator(@"
+                using System.Text.RegularExpressions;
+                public partial class A
+                {
+                    public partial record class B
+                    {
+                        public partial record struct C
+                        {
+                            public partial record D
+                            {
+                                public partial struct E
+                                {
+                                    [RegexGenerator(""ab"")]
+                                    public static partial Regex Valid();
+                                }
+                            }
+                        }
+                    }
+                }
+            ", compile: true));
+        }
```

Comment:
Consider testing a local function as well

Reply:
The methods need to be partial.  Partial locals aren't a thing.

Reply:
Yeah I meant that it produces a reasonable diagnostic, not that it compiles

Reply:
But it's not even valid syntax.

Reply:
@stephentoub Let me be more clear here, the attribute will be added on a non-partial local function, which is a valid syntax.

Currently, there is:

```csharp
                        // Scope to just methods
                        if (member is not MethodDeclarationSyntax methodSyntax)
                        {
                            continue;
                        }
```

which seems to silently ignore the attribute for that case. My expectation is to produce a diagnostic for that. Or if you want to silently skip it, then it's at least tested.

Reply:
Thanks, but I still don't understand why a generator should have to care about that at all.  Look how many errors you get if you even try to do this:
![image](https://user-images.githubusercontent.com/2642209/133617248-aafafa22-2866-4a4f-b951-a758d3699071.png)

Reply:
> the attribute will be added on a non-partial local function, which is a valid syntax.

You mean if the dev puts it on a non-partial local function?  I see; I can make sure the existing diagnostics fire correctly for that.

Reply:
@stephentoub I'm referring to a non-partial local function, which compiles fine:

```csharp
using System;
using System.Text.RegularExpressions;

[AttributeUsage(AttributeTargets.Method, AllowMultiple = false)]
public class MyAttribute : Attribute
{
}

public class C
{
    public void M()
    {
        Abc();

        [My]
        static Regex Abc() => null;        
    }
}
```

![image](https://user-images.githubusercontent.com/31348972/133617929-6b914748-0849-4849-a137-1e8a160a3799.png)

Do we want this to be just skipped as if the attribute isn't there? or do we want a diagnostic? whatever the desired behavior, it's good to be tested. I know this case isn't really important and both behaviors could be reasonable. But just for sake of completeness.

Reply:
> Do we want this to be just skipped as if the attribute isn't there?

The loop you cite won't be skipping local functions, as it's iterating over members, which won't ever contain local functions.  We'd need to scour the whole syntax tree looking for local functions, in which case it seems like we'd need to essentially change IsSyntaxTargetForGeneration/GetSemanticTargetForGeneration to essentially return every method or property and then have GetRegexTypesToEmit look for any local functions that are attributed.  Is that what you're proposing be done?  Note we don't have the ability to report a diagnostic in GetSemanticTargetForGeneration, nor do I believe we have the ability to cache things like type symbols from the compilation.